### PR TITLE
Implementing Limb Darkening correction

### PR DESCRIPTION
**Pull request by Raahul-Singh**
This is in continuation of #2344
I have refactored the code. This requires testing, though I couldn't find any appropriate test data.
Test data suggestions are welcome!

### CODE REVIEW

File: changelog/3728.feature.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Adds Limb Darkening Effect correction feature for `~sunpy.map.GenericMap`
```
Comment (nabobalis) [inline changelog/3728.feature.rst]:
```suggestion
Adds a limb darkening correction function (`sunpy.physics.get_limb_dark`) that can be used on sunpy maps.
```

File: sunpy/physics/limb_darkening.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    Limitations
+    -----------
+    Only applies to wavelengths in the range 4000<lambda<15000 Angstrom
+
+    Reference
+    ---------
+    Cox, Arthur N., Allen's astrophysical quantities, 4th ed. Publisher:
+        New York: AIP Press; Springer, 2000. ISBN: 038798746
+
```
Comment (nabobalis) [inline sunpy/physics/limb_darkening.py]:
```suggestion
```

File: sunpy/physics/limb_darkening.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    Quantities by Allen.
+
+    Limitations
+    -----------
+    Only applies to wavelengths in the range 4000<lambda<15000 Angstrom
+
+    Reference
+    ---------
+    Cox, Arthur N., Allen's astrophysical quantities, 4th ed. Publisher:
+        New York: AIP Press; Springer, 2000. ISBN: 038798746
```
Comment (nabobalis) [inline sunpy/physics/limb_darkening.py]: I wonder if this has a DOI we can replace this with.

File: sunpy/physics/limb_darkening.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    Notes
+    -----
+    Coefficients for 5th order polynomial fit to limb darkening
+    constant `u` was done on data from Astrophysical
+    Quantities by Allen.
+
+    Limitations
+    -----------
+    Only applies to wavelengths in the range 4000<lambda<15000 Angstrom
```
Comment (nabobalis) [inline sunpy/physics/limb_darkening.py]: I am not sure if theses should be sections and not just part of the first paragraph.

File: sunpy/physics/limb_darkening.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    Reference
+    ---------
```
Comment (nabobalis) [inline sunpy/physics/limb_darkening.py]:
```suggestion
    ----------
```

File: sunpy/physics/limb_darkening.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    -----
+    Coefficients for 5th order polynomial fit to limb darkening
+    constant `u` was done on data from Astrophysical
+    Quantities by Allen.
+
+    Limitations
+    -----------
+    Only applies to wavelengths in the range 4000<lambda<15000 Angstrom
+
+    Reference
```
Comment (nabobalis) [inline sunpy/physics/limb_darkening.py]:
```suggestion
    References
```

File: sunpy/physics/limb_darkening.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    """
+
+    au = -8.9829751
+    bu = 0.0069093916
+    cu = -1.8144591e-6
+    du = 2.2540875e-10
+    eu = -1.3389747e-14
+    fu = 3.0453572e-19
+
+    a = np.array((au, bu, cu, du, eu, fu))
```
Comment (nabobalis) [inline sunpy/physics/limb_darkening.py]: What are these?

File: sunpy/physics/limb_darkening.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+    wavelength = wavelength.to_value()
+    wavelength_series = np.array([wavelength ** n for n in range(len(a))])
+
+    vl = np.sum(a * wavelength_series)
+
+    return vl
+
+
+def get_limb_dark(sun_map):
```
Comment (nabobalis) [inline sunpy/physics/limb_darkening.py]: I never did like the name of this but I don't have any other ideas.

File: sunpy/physics/limb_darkening.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    Only applies to wavelengths in the range 4000<lambda<15000 Angstrom
+
+    Examples
+    --------
+    >>> import sunpy.map
+    >>> from sunpy.phyisics import limb_darkening
+    >>> mymap = sunpy.map.Map('test.fits')
+    >>> newmap = limb_darkening.correct_limb_darkening(mymap)
+    >>> mymap.peek()
+    >>> newmap.peek()
```
Comment (nabobalis) [inline sunpy/physics/limb_darkening.py]: Remove these two lines, we can't peek in examples if i recall.

File: sunpy/physics/limb_darkening.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+    ul = darklimb_u(sun_map.wavelength)
+    vl = darklimb_v(sun_map.wavelength)
+
+    x_center, y_center = sun_map.world_to_pixel(SkyCoord(0, 0, unit=u.arcsec, frame=sun_map.coordinate_frame))
+    x_center = x_center.to(u.pixel)
+    y_center = y_center.to(u.pixel)
+    x_dim = sun_map.dimensions.x.to(u.pixel).value
+    y_dim = sun_map.dimensions.y.to(u.pixel).value
+    radius = (sun_map.rsun_obs.to(u.arcsec).value /
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]: This whole section seems way too happy with de-uniting things. This radius calculation in particular doesn't need to do a `to_value` until right at the end.

I would be more interested in knowing if this whole algorithm could be made to be unitful all the way through, i.e. to `to_value` anywhere.

File: sunpy/physics/limb_darkening.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    --------
+    >>> import sunpy.map
+    >>> from sunpy.phyisics import limb_darkening
+    >>> mymap = sunpy.map.Map('test.fits')
+    >>> newmap = limb_darkening.correct_limb_darkening(mymap)
+    >>> mymap.peek()
+    >>> newmap.peek()
+    """
+    new_data = sun_map.data / get_limb_dark(sun_map)
+    new_meta = sun_map.meta
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]: This is updating the old meta inplace as it is at the moment.
```suggestion
    new_meta = copy.deepcopy(sun_map.meta)
```

File: sunpy/physics/limb_darkening.py (hunk 8 above)
Comment (Cadair) [reply sunpy/physics/limb_darkening.py]: I think if we make all these private we need to worry less.

File: sunpy/physics/limb_darkening.py (hunk 7 above)
Comment (Raahul-Singh) [reply sunpy/physics/limb_darkening.py]: To model the limb darkening effect, the limb darkening function uses a 5th order polynomial fitting to the limb darkening constants obtained from Astrophysical Quantities.
These are the six constants taken from Allen's book.
For reference : [The IDL algorithm](https://hesperia.gsfc.nasa.gov/ssw/gen/idl/solar/darklimb_correct.prol)

File: sunpy/physics/limb_darkening.py (hunk 9 above)
Comment (Raahul-Singh) [reply sunpy/physics/limb_darkening.py]: Done

File: sunpy/physics/limb_darkening.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
+import numpy as np
+from astropy import units as u
+from astropy.coordinates import SkyCoord
+
+
+def darklimb_u(wavelength):
```
[the remark this answers is not in the document]
Comment (Raahul-Singh) [reply sunpy/physics/limb_darkening.py]: Done

**Comment by Raahul-Singh** [comment]
I need data to add more tests.

**Comment by nabobalis** [comment]
Is it not possible in the mean time to create artificial tests?

File: sunpy/physics/limb_darkening.py [hunk 13]
```diff
@@ -38,39 +36,33 @@ def darklimb_u(wavelength):
...
+    wavelength_series = np.array([wavelength.value ** n for n in range(len(a))])
 
     ul = np.sum(a * wavelength_series)
 
     return ul
 
-
-def darklimb_v(wavelength):
+@u.quantity_input
+def __darklimb_v(wavelength):
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]:
```suggestion
def _darklimb_v(wavelength: u.AA):
```

File: sunpy/physics/limb_darkening.py [hunk 14]
```diff
@@ -82,15 +74,14 @@ def darklimb_v(wavelength):
 
     a = np.array((av, bv, cv, dv, ev, fv))
 
-    wavelength = wavelength.to_value()
-    wavelength_series = np.array([wavelength ** n for n in range(len(a))])
+    wavelength_series = np.array([wavelength.value ** n for n in range(len(a))])
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]: Do we need to `.value` here or can we do this whole function with units?

File: sunpy/physics/limb_darkening.py [hunk 15]
```diff
@@ -82,15 +74,14 @@ def darklimb_v(wavelength):
...
 
     vl = np.sum(a * wavelength_series)
 
     return vl
 
 
-def get_limb_dark(sun_map):
+def __get_limb_dark(sun_map):
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]:
```suggestion
def _get_limb_dark(sun_map):
```

File: sunpy/physics/limb_darkening.py [hunk 16]
```diff
@@ -117,30 +108,29 @@ def get_limb_dark(sun_map):
...
 
     if not (4000 * u.AA < sun_map.wavelength < 15000 * u.AA):
-        raise ValueError("The wavelength of the observation \
-                          must be between 4000 and 15000 Angstrom.")
+        raise ValueError("The wavelength of the observation " +
+                          "must be between 4000 and 15000 Angstrom.")
 
-    ul = darklimb_u(sun_map.wavelength)
-    vl = darklimb_v(sun_map.wavelength)
+    ul = __darklimb_u(sun_map.wavelength)
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]:
```suggestion
    ul = _darklimb_u(sun_map.wavelength)
```

File: sunpy/physics/limb_darkening.py [hunk 17]
```diff
@@ -117,30 +108,29 @@ def get_limb_dark(sun_map):
...
+    vl = __darklimb_v(sun_map.wavelength)
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]:
```suggestion
    vl = _darklimb_v(sun_map.wavelength)
```

File: sunpy/physics/limb_darkening.py [hunk 18]
```diff
@@ -172,12 +162,10 @@ def correct_limb_darkening(sun_map):
     >>> from sunpy.phyisics import limb_darkening
     >>> mymap = sunpy.map.Map('test.fits')
     >>> newmap = limb_darkening.correct_limb_darkening(mymap)
-    >>> mymap.peek()
-    >>> newmap.peek()
     """
-    new_data = sun_map.data / get_limb_dark(sun_map)
+    new_data = sun_map.data / __get_limb_dark(sun_map)
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]:
```suggestion
    new_data = sun_map.data / _get_limb_dark(sun_map)
```

File: sunpy/physics/tests/test_limb_darkening.py [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+
+@pytest.fixture
+def aia171_test_map():
+    testpath = sunpy.data.test.rootdir
+    return sunpy.map.Map(os.path.join(testpath, 'aia_171_level1.fits'))
+
+def test_limb_filter(aia171_test_map):
+    with pytest.raises(ValueError):
+        limb_darkening.__get_limb_dark(aia171_test_map)
```
Comment (Cadair) [inline sunpy/physics/tests/test_limb_darkening.py]:
```suggestion
        limb_darkening._get_limb_dark(aia171_test_map)
```

File: sunpy/physics/limb_darkening.py [hunk 20]
```diff
@@ -117,30 +108,29 @@ def get_limb_dark(sun_map):
...
-    radius = (sun_map.rsun_obs.to(u.arcsec).value /
-              sun_map.scale.axis1.to(u.arcsec / u.pixel).value)
+    x_center = x_center
+    y_center = y_center
+    x_dim = sun_map.dimensions.x
+    y_dim = sun_map.dimensions.y
+    radius = sun_map.rsun_obs / sun_map.scale.axis1
 
-    x_grid, y_grid = np.meshgrid(np.arange(int(x_dim)), np.arange(int(y_dim)), sparse=True)
+    x_grid, y_grid = np.meshgrid(np.arange(x_dim.value), np.arange(y_dim.value), sparse=True)
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]: You should always use `to_value` to be explicit about what units you are dropping.

File: sunpy/physics/limb_darkening.py [hunk 21]
```diff
@@ -117,30 +108,29 @@ def get_limb_dark(sun_map):
...
+    vl = __darklimb_v(sun_map.wavelength)
 
     x_center, y_center = sun_map.world_to_pixel(SkyCoord(0, 0, unit=u.arcsec, frame=sun_map.coordinate_frame))
-    x_center = x_center.to(u.pixel)
-    y_center = y_center.to(u.pixel)
-    x_dim = sun_map.dimensions.x.to(u.pixel).value
-    y_dim = sun_map.dimensions.y.to(u.pixel).value
-    radius = (sun_map.rsun_obs.to(u.arcsec).value /
-              sun_map.scale.axis1.to(u.arcsec / u.pixel).value)
+    x_center = x_center
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]:
```suggestion
```

File: sunpy/physics/limb_darkening.py [hunk 22]
```diff
@@ -117,30 +108,29 @@ def get_limb_dark(sun_map):
...
+    y_center = y_center
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]:
```suggestion
```

File: sunpy/physics/limb_darkening.py [hunk 23]
```diff
@@ -1,32 +1,30 @@
 import numpy as np
 from astropy import units as u
 from astropy.coordinates import SkyCoord
+import copy
 
 
-def darklimb_u(wavelength):
+@u.quantity_input
+def __darklimb_u(wavelength):
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]: We should only be using a single underscore to indicate "private", also you need to annotate the argument with the expected unit:
```suggestion
def _darklimb_u(wavelength: u.AA):
```

File: sunpy/physics/limb_darkening.py [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+
+
+@u.quantity_input
+def __darklimb_u(wavelength):
+    """
+    Returns the constant, u, for a given wavelength for input into the limb
+    darkening function.
+
+    Coefficients for 5th order polynomial fit to limb darkening
+    constant `u` was done on data from Astrophysical Quantities by Allen.
```
Comment (Cadair) [inline sunpy/physics/limb_darkening.py]:
```suggestion
    constant ``u`` was done on data from Astrophysical Quantities by Allen.
```

File: sunpy/physics/limb_darkening.py (hunk 14 above)
Comment (Raahul-Singh) [reply sunpy/physics/limb_darkening.py]: It would only work with scalars. Otherwise throws an error.

File: sunpy/physics/limb_darkening.py (hunk 21 above)
Comment (Raahul-Singh) [reply sunpy/physics/limb_darkening.py]: I forgot to delete these :man_facepalming:

**Comment by Raahul-Singh** [comment]
> Is it not possible in the mean time to create artificial tests? Can you also fix the failing CI builds?

Is there a way to get a sunmap with the wavelength in the supported range? Otherwise, I don't have enough solar physics knowledge to create a theoretical test. :sweat_smile:

**Comment by nabobalis** [comment]
> > Is it not possible in the mean time to create artificial tests? Can you also fix the failing CI builds?
> 
> Is there a way to get a sunmap with the wavelength in the supported range? Otherwise, I don't have enough solar physics knowledge to create a theoretical test.

Some of the test data we have should have the correct wavelength. HMI continuum or SDO/AIA 4500.

When I say theoretical here, I just mean, create a fake map with some numbers and see what happens.

**Comment by Raahul-Singh** [comment]
> > > Is it not possible in the mean time to create artificial tests? Can you also fix the failing CI builds?
> > 
> > 
> > Is there a way to get a sunmap with the wavelength in the supported range? Otherwise, I don't have enough solar physics knowledge to create a theoretical test.
> 
> Some of the test data we have should have the correct wavelength. HMI continuum or SDO/AIA 4500.
> 
> When I say theoretical here, I just mean, create a fake map with some numbers and see what happens.


So I tried downloading some data, ran the code and this was the result.:
map :
before
map with `limb darkening` removed :
after

**Comment by nabobalis** [comment]
Looks pretty good.

**Comment by Raahul-Singh** [comment]
What should the tests be?

**Comment by nabobalis** [comment]
> What should the tests be?
All I can think of is: one figure test and ideally some sort of artificial numerical test

**Comment by Raahul-Singh** [comment]
> > What should the tests be?
> > All I can think of is: one figure test and ideally some sort of artificial numerical test

For a figure test, I'd need some standard data. As for the numerical test, I am kinda blank here. The only thing I can think of is to test the `limb filter` matrix that is generated for a given wavelength against known data or some mathematical expression that should yield that matrix for that particular wavelength.

**Comment by nabobalis** [comment]
I don't really have answers to those questions. Hopefully others too.

**Comment by Raahul-Singh** [comment]
I found a pair of intensitygrams with and without limb darkening from [stanford/soi/data](http://soi.stanford.edu/data/).

* With limb darkening

MDI_int_1999 02 07_12 00 rotated_with_limb_darkening

* With limb darkening removed

MDI_int_1999 02 07_12 00 rotated


The problem is that there are in `gif` format. How should I go about making maps out of them?

File: sunpy/physics/limb_darkening.py [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+
+    dist_grid = np.sqrt(x_2 + y_2) / radius.value
+    dist_grid[np.where(dist_grid > 1)] = 0
+
+    limbfilt = 1 - ul - vl + ul * np.cos(np.arcsin(dist_grid)) + vl * np.cos(np.arcsin(dist_grid)) ** 2
+
+    return limbfilt
+
+
+def correct_limb_darkening(sun_map):
```
Comment (dstansby) [inline sunpy/physics/limb_darkening.py:139]: Thanks for the PR! I think this should have a reference for the method, or failing that a detailed mathematical description of exactly what the method does.

**Comment by nabobalis** [comment]
I will be closing this, we need someone to pick this up with background on this topic.

Thank you @Raahul-Singh for the effort you put into this PR.

**Comment by AKMourato** [comment]
Greetings.
Is there any functional implementation on this matter?

**Comment by dstansby** [comment]
There is not I'm afraid. If you would like to pick this up and implement it please do let us know and we can help.

**Comment by Cadair** [comment]
Thanks for your interest @AKMourato the canonical issue tracking this is https://github.com/sunpy/sunkit-image/issues/68 if you are interested on working on this we would love to help you :smile:

**Pull request closed without merging**